File: LawFlow/frontend/src/api/tutor.ts
import api from "@/lib/api";
import type { TutorSession, TutorMode } from "@/types";

export async function getModes(): Promise<Record<string, TutorMode>> {
  const { data } = await api.get("/tutor/modes");
  return data;
}

export async function createSession(params: {
  mode: string;
  subject?: string;
  topics?: string[];
}): Promise<TutorSession> {
  const { data } = await api.post("/tutor/session", params);
  return data;
}

export async function getSession(sessionId: string): Promise<TutorSession> {
  const { data } = await api.get(`/tutor/session/${sessionId}`);
  return data;
}

export async function endSession(sessionId: string): Promise<TutorSession> {
  const { data } = await api.post(`/tutor/session/${sessionId}/end`);
  return data;
}

export async function sendMessageStream(
  sessionId: string,
  content: string,
  onChunk: (text: string) => void,
  onDone: () => void
): Promise<void> {
  const response = await fetch("/api/tutor/message", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ session_id: sessionId, content }),
  });

  if (!response.ok) {
    throw new Error(`Tutor API error: ${response.status}`);
  }

  const reader = response.body?.getReader();
  if (!reader) throw new Error("No response body");

  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() || "";

    for (const line of lines) {
      if (line.startsWith("data: ")) {
        const text = line.slice(6);
        if (text === "[DONE]") {
          onDone();
          return;
        }
        if (text.startsWith("[ERROR]")) {
          throw new Error(text);
        }
        onChunk(text);
      }
    }
  }
  onDone();
}

File: LawFlow/api/routes/tutor.py
"""AI tutor session routes with SSE streaming."""

from flask import Blueprint, request, jsonify, Response

from api.errors import ValidationError, NotFoundError
from api.services import tutor_engine
from api.services.prompt_library import MODES

bp = Blueprint("tutor", __name__, url_prefix="/api/tutor")


@bp.route("/modes", methods=["GET"])
def list_modes():
    """List available tutor modes."""
    mode_info = {
        "socratic": {"name": "Socratic Questioning", "description": "Learn through guided questions that probe your understanding"},
        "irac": {"name": "IRAC Practice", "description": "Practice structured legal analysis: Issue, Rule, Application, Conclusion"},
        "issue_spot": {"name": "Issue Spotting", "description": "Train to identify all legal issues in complex fact patterns"},
        "hypo": {"name": "Hypothetical Drill", "description": "Test rule boundaries by modifying facts and analyzing changes"},
        "explain": {"name": "Explain (Catch Up)", "description": "Compressed, high-signal teaching for rapid concept mastery"},
        "exam_strategy": {"name": "Exam Strategy", "description": "Master exam technique, time management, and answer structure"},
    }
    return jsonify(mode_info)


@bp.route("/session", methods=["POST"])
def create_session():
    """Start a new tutoring session."""
    data = request.get_json()
    if not data:
        raise ValidationError("JSON body required")

    mode = data.get("mode", "explain")
    if mode not in MODES:
        raise ValidationError(f"Invalid mode. Available: {', '.join(MODES.keys())}")

    session = tutor_engine.create_session(
        mode=mode,
        subject=data.get("subject"),
        topics=data.get("topics"),
    )
    return jsonify(session), 201


@bp.route("/session/<session_id>", methods=["GET"])
def get_session(session_id: str):
    """Get session details with message history."""
    session = tutor_engine.get_session(session_id)
    if not session:
        raise NotFoundError("Session not found")
    return jsonify(session)


@bp.route("/message", methods=["POST"])
def send_message():
    """Send a message and stream Claude's response via SSE."""
    data = request.get_json()
    if not data:
        raise ValidationError("JSON body required")

    session_id = data.get("session_id")
    content = data.get("content")

    if not session_id or not content:
        raise ValidationError("session_id and content are required")

    def generate():
        try:
            for chunk in tutor_engine.send_message(session_id, content):
                yield f"data: {chunk}\n\n"
            yield "data: [DONE]\n\n"
        except ValueError as e:
            yield f"data: [ERROR] {str(e)}\n\n"

    return Response(generate(), mimetype="text/event-stream")


@bp.route("/session/<session_id>/end", methods=["POST"])
def end_session(session_id: str):
    """End a tutoring session."""
    result = tutor_engine.end_session(session_id)
    if not result:
        raise NotFoundError("Session not found")
    return jsonify(result)
